akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,363 @@
+      log.debug("Received Journal response: {}", response)
+      response match {
+        case WriteMessageSuccess(p, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case we ignore the call to the handler
+          if (id == writerIdentity.instanceId) {
+            updateLastSequenceNr(p)
+            popApplyHandler(p.payload)
+            onWriteMessageComplete()
+            tryUnstash(context, applySideEffects(sideEffects))
+          } else same
+
+        case WriteMessageRejected(p, cause, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case the handler has already been discarded
+          if (id == writerIdentity.instanceId) {
+            updateLastSequenceNr(p)
+            onPersistRejected(cause, p.payload, p.sequenceNr) // does not stop
+            tryUnstash(context, applySideEffects(sideEffects))
+          } else same
+
+        case WriteMessageFailure(p, cause, id) ⇒
+          // instanceId mismatch can happen for persistAsync and defer in case of actor restart
+          // while message is in flight, in that case the handler has already been discarded
+          if (id == writerIdentity.instanceId) {
+            onWriteMessageComplete()
+            onPersistFailureThenStop(cause, p.payload, p.sequenceNr)
+          } else same
+
+        case WriteMessagesSuccessful ⇒
+          // ignore
+          same
+
+        case WriteMessagesFailed(_) ⇒
+          // ignore
+          same // it will be stopped by the first WriteMessageFailure message; not applying side effects
+
+        case _: LoopMessageSuccess ⇒
+          // ignore, should never happen as there is no persistAsync in typed
+          same
+      }
+    }
+
+    private def onWriteMessageComplete(): Unit =
+      tryBecomeHandlingCommands()
+
+    private def onPersistRejected(cause: Throwable, event: Any, seqNr: Long): Unit = {
+      log.error(
+        cause,
+        "Rejected to persist event type [{}] with sequence number [{}] for persistenceId [{}] due to [{}].",
+        event.getClass.getName, seqNr, persistenceId, cause.getMessage)
+    }
+
+    private def onPersistFailureThenStop(cause: Throwable, event: Any, seqNr: Long): Behavior[Any] = {
+      log.error(cause, "Failed to persist event type [{}] with sequence number [{}] for persistenceId [{}].",
+        event.getClass.getName, seqNr, persistenceId)
+
+      // FIXME see #24479 for reconsidering the stopping behaviour
+      Behaviors.stopped
+    }
+
+  }
+
+  // the active phase switches between PersistingEvents and HandlingCommands;
+  // we do this via a var instead of behaviours to keep allocations down as this will be flip/flaping on every Persist effect
+  private[this] var phase: EventsourcedRunningPhase = HandlingCommands
```

Comment:
will be interesting when this is changed to immutable, I assume that the different phases will simply be different behaviors